Merge 2.12.x to 2.13.x [ci: last-only]

retronym opened the pull request:
```
% git co origin/2.13.x; export MB=$(git merge-base head origin/2.12.x); git log --graph --oneline $MB..origin/2.12.x | tail -r
HEAD is now at 4bfe574938 Merge pull request #7128 from retronym/perf/find-dir-entry
* ab99db089b Rewrite List(a, b) to `new ::(a, new ::(b, Nil)`
* 296e3167a8 add a simple test for optimised List.apply rewrite in cleanup
* 9a2219405b Merge pull request #6994 from rorygraves/mike/2.12.x_ListApply
| * 0763168ca6 [nomerge] Emit bridge method forwarders with BRIDGE flag
|\
*   4c9780fb7a Merge pull request #7035 from eed3si9n/bport/b10812
|/
| * 49a5ebde96 Uncurry of Java varargs respects Java's universal trait nescience
|\
*   15f75ad98d Merge pull request #7141 from hrhino/t11109
| * 42dac30eb5 [nomerge] fix performance regression in mutable.HashMap#getOrElseUpdate
|\
*   2b2356e70f Merge pull request #7091 from SethTisue/mutable-hashmap-getorelseupdate
| * aa7578210b Don't re-apply type maps to info of unchanged symbols
| * 228d21780d Make internal methods in TypeMap more private/final/inlinable
|\
*   38282c7af1 Merge pull request #7137 from hrhino/faster/type-map

% git merge 9a2219405b
Auto-merging src/reflect/scala/reflect/internal/Definitions.scala
CONFLICT (content): Merge conflict in src/reflect/scala/reflect/internal/Definitions.scala
Auto-merging src/compiler/scala/tools/nsc/transform/CleanUp.scala
CONFLICT (content): Merge conflict in src/compiler/scala/tools/nsc/transform/CleanUp.scala
Automatic merge failed; fix conflicts and then commit the result.

% git merge -sours 4c9780fb7a
Merge made by the 'ours' strategy.

% git merge -sours 2b2356e70f
Merge made by the 'ours' strategy.

% git merge 38282c7af1
Auto-merging src/reflect/scala/reflect/internal/tpe/TypeMaps.scala
CONFLICT (content): Merge conflict in src/reflect/scala/reflect/internal/tpe/TypeMaps.scala
Automatic merge failed; fix conflicts and then commit the result.
```

retronym committed:
Rewrite List(a, b) to `new ::(a, new ::(b, Nil)`

Conservatively limits the extra stack frames consumed by the generated
program to 8.

 Author:    Jason Zaugg <jzaugg@gmail.com>
 Date:      Wed Aug 1 00:02:59 2018 +0100

eed3si9n committed:
[nomerge] Emit bridge method forwarders with BRIDGE flag

Fixes scala/bug#11061
Ref scala/bug#10812

On 2.13.x branch #6531 removed the mirror class forwarders for bridge methods. I would like to do same in 2.12.x since Java 11-ea started to find them ambiguous as seen in akka/akka#25449 / scala/bug#11061.

To keep binary compatibility, I am still emitting the forwarders for bridge methods, but with `ACC_BRIDGE` flag.

mkeskells committed:
add a simple test for optimised List.apply rewrite in cleanup

SethTisue committed:
[nomerge] fix performance regression in mutable.HashMap#getOrElseUpdate

the change in question originated in
https://github.com/scala/collection-strawman/pull/484. it was correct
at the time because `HashTable#addEntry0` has a threshold check

but then when the change was backported to 2.12.x in
https://github.com/scala/scala/pull/6828, the
`HashTable#addEntry0`call was replaced with a call to
`HashMap#addEntry0`, which doesn't check the threshold. so if the
table is only ever updated using `getOrElseUpdate`, the table's load
factor would just keep climbing, resulting in poor performance

this was caught by my Project Euler solutions :-)

[nomerge] since the problem is specific to the 2.12 code

dwijnand committed:
Merge pull request #6994 from rorygraves/mike/2.12.x_ListApply

Rewrite List(a, b) to `new ::(a, new ::(b, Nil)`

lrytz committed:
Merge pull request #7035 from eed3si9n/bport/b10812

Emit bridge method forwarders with BRIDGE flag

hrhino committed:
Don't re-apply type maps to info of unchanged symbols

The fast path in `mapOver` would avoid cloning symbols if the map
had no effect on any of their infos. However, if a symbol with a
changed info was found, it would use `cloneSymbolsAndModify` to
apply itself to a fresh clone of those symbols, with the result
that it would re-apply itself to the infos of symbols clone from
symbols that were unchanged.

This avoids that.

In library/reflect/compiler, this avoids 51358 _direct_ repeated
calls to `TypeMap#apply`. Since those may go off and do more work,
it's more than that.

hrhino committed:
Uncurry of Java varargs respects Java's universal trait nescience

If `U` is a universal trait (i.e., one extending from `Any` rather
than `AnyRef`), Java still sees it as a normal interface, and in
a method with a generic varargs parameter `<T extends U>`, interprets
that as declaring an array argument `U[]` after erasure. However,
due to a pre-value-class implementation of `isUnboundedGeneric`,
such a type parameter was considered unbounded, and therefore Scala
considered its erasure to be `Object[]` instead. This causes a runtime
`NoSuchMethodError`, of course.

I moved `isUnboundedGeneric` from `Types` into `UnCurry`, since that's
the only place it was used, and its proximity to `isBoundedGeneric`,
which is neither defined as nor synonymous with `!isUnboundedGeneric`
appeared likely to cause confusion. (_That_ method is only used in
`SpecialiseTypes`, but I didn't want to change an unrelated file.)

Running into this bug is probably penance for using Hibernate with
Scala, but I promise I'm "just following orders".

Fixes scala/bug#11109.

lrytz committed:
Merge pull request #7141 from hrhino/t11109

Uncurry of Java varargs respects Java's universal trait nescience

retronym committed:
Merge pull request #7091 from SethTisue/mutable-hashmap-getorelseupdate

fix performance regression in mutable.HashMap#getOrElseUpdate

retronym committed:
Make internal methods in TypeMap more private/final/inlinable

retronym committed:
Merge pull request #7137 from hrhino/faster/type-map

Don't re-apply type maps to info of unchanged symbols

retronym committed:
Merge commit '9a2219405b' into HEAD

retronym committed:
Merge commit '4c9780fb7a' into HEAD

retronym committed:
Merge commit '2b2356e70f' into HEAD

retronym committed:
Merge commit '38282c7af1' into HEAD

lrytz commented:
```
sbt.ForkMain$ForkError: java.lang.ClassCastException: scala.collection.immutable.$colon$colon cannot be cast to scala.collection.immutable.Set
	at scala.reflect.internal.StdNames$TermNames.<init>(StdNames.scala:952)
```

retronym committed:
Fix semantic merge errors

  - List.apply in inherited in 2.13.x, so adjust the intrinsic to check the
    qualifier from which it is selected.
  - Doing this seems to have exposed a latent bug in the unpickler. When
    ListModule is completed before ListClass, only the module gets
    associatedFile set. This turned into a test failure in run/t191.scala
    I've changed the type completer to set both the class and module
    symbols's associated file instead.
  - Avoid procedure syntax deprecation in a new test.

retronym commented:
Pushed a commit to fix the semantic merge error.

SethTisue commented:
```
[test] !!  235 - run/list-apply-eval.scala                 [output differs]
[test] % diff /home/jenkins/workspace/scala-2.13.x-validate-main/test/files/run/list-apply-eval.check /home/jenkins/workspace/scala-2.13.x-validate-main/test/files/run/list-apply-eval-run.log--- empty
[test] +++ list-apply-eval-run.log
[test] @@ -1,0 +1,1 @@
[test] +warning: there was one deprecation warning (since 2.13.0); re-run with -deprecation for details
[test] ##### Log file '/home/jenkins/workspace/scala-2.13.x-validate-main/test/files/run/list-apply-eval-run.log' from failed test #####
[test] 
[test] warning: there was one deprecation warning (since 2.13.0); re-run with -deprecation for details
[test] 
```

```
[test] !! 26 - run/t191.scala                            [output differs][duration 5.56s]
[test] % diff /home/jenkins/workspace/scala-2.13.x-validate-main/test/scaladoc/run/t191.check /home/jenkins/workspace/scala-2.13.x-validate-main/test/scaladoc/run/t191-run.log
[test] @@ -1 +1,48 @@
[test] -Done.
[test] +java.lang.AssertionError: assertion failed
```

lrytz commented on src/reflect/scala/reflect/internal/Definitions.scala:
```diff
@@ -1587,7 +1587,21 @@ trait Definitions extends api.StandardDefinitions {
 
       lazy val Option_apply = getMemberMethod(OptionModule, nme.apply)
-      lazy val List_apply = DefinitionsClass.this.List_apply
+      private lazy val List_apply = DefinitionsClass.this.List_apply
```
The field name `List_apply` seems quite misleading, should we rename it to `IterableFactory` (and also resolve as such)?

lrytz reviewed (approved):
LGTM, we can fix the field name post factum

lrytz commented:
Argh, it seems https://github.com/scala/scala/pull/7141 got merged with `-s ours` by mistake - or did I miss something?
